apps/app-loader/spawn/src/main.rs

```diff
@@ -0,0 +1,161 @@
+    if let Some(msg) = envelope.body.scalar_message() {
+        if envelope.body.is_blocking() {
+            xous::syscall::return_scalar(envelope.sender, msg.arg1 + 1).unwrap();
+        }
+    }
+}
+
+fn read_elf(memory: Option<&mut xous::MemoryMessage>) -> usize {
+    let memory = match memory {
+        Some(s) => s,
+        None => panic!(),
+    };
+
+    // get the elf binary from the message
+    let mut bin = memory.buf.as_slice::<u8>();
+
+    // go to the beginning of the ELF file using the provided offset
+    bin = &bin[memory.offset.and_then(|n| Some(n.get())).unwrap_or(0)..];
+
+    let max = bin.len();
+
+    // make sure that this is a 32 bit
+    assert!(bin[4] == 0x01);
+
+    // a helper function to get a region of the file as a usize
+    let to_usize = |start, size| {
+	if size == 1 {
+	    return bin[start] as usize;
+	}
+	if size == 2 {
+	    // assumes little endianness
+	    return u16::from_le_bytes(bin[start..start+size].try_into().unwrap()) as usize;
+	}
+	if size == 4 {
+	    return u32::from_le_bytes(bin[start..start+size].try_into().unwrap()) as usize;
+	}
+	panic!("Tried to get usize of invalid size!");
+    };
+
+    // some basic stuff to know
+    let entry_point = to_usize(0x18, 4);
+    let ph_start = to_usize(0x1c, 4);
+    let ph_size = to_usize(0x2A, 2);
+    let ph_count = to_usize(0x2C, 2);
+
+    // add the segments we should load
+    for i in 0..ph_count {
+	let start = ph_start + i * ph_size;
+	// only load PT_LOAD segments
+	if  to_usize(start, 4) == 0x00000001 {
+	    let src_addr = to_usize(start+0x04, 4);
+	    let vaddr = to_usize(start+0x08, 4);
+	    let padding = if vaddr & 0xFFF == 0 { 0 } else { vaddr & 0xFFF };
+	    let file_size = to_usize(start+0x10, 4);
+	    let mem_size = to_usize(start+0x14, 4);
+	    let mem_size = mem_size + padding;
+	    let mem_size = mem_size + if mem_size & 0xFFF == 0 { 0 } else { 0x1000 - (mem_size & 0xFFF) };
+
+	    assert_eq!(0, mem_size & 0xFFF);
+	    assert_eq!(0, (vaddr - padding) & 0xFFF);
+
+	    log::info!("Loading offset {} to virtual address {} with memory size {}", src_addr, vaddr, mem_size);
+	    let mut target_memory = xous::map_memory(
+		None,
+		core::num::NonZeroUsize::new(vaddr-padding),
+		mem_size,
+		xous::MemoryFlags::R | xous::MemoryFlags::W | xous::MemoryFlags::X,
+	    )
+		.unwrap();
+
+	    for (dest, src) in target_memory.as_slice_mut().iter_mut().skip(padding)
+		.zip(bin[src_addr..core::cmp::min(max, src_addr+file_size)].iter())
+	    {
+		*dest = *src;
+	    }
+	    for dest in target_memory.as_slice_mut().iter_mut().skip(padding+file_size)
+		.take(mem_size-file_size-padding)
+	    {
+		*dest = 0;
+	    }
+	}
+    }
+
+    memory.offset = None;
+    log::info!("Finished writing");
+    return entry_point;
+}
+
+fn jump(entry_point: usize) -> ! {
+    log::info!("Jumping to {}", entry_point);
+    let entry_fn = unsafe { core::mem::transmute::<_, fn() -> !>(entry_point as *const u8) };
```

Comment:
A future version of Xous will add the `.eh_frame` offset as the first argument, and may add a thread-local-storage pointer and length as the second and third arguments. No need to update this, but it's something to be aware of. We'll get unwinding support soon as a result of this change.